# Data Science Unit 1 Sprint Challenge 4

## Exploring Data, Testing Hypotheses

In this sprint challenge you will look at a dataset of people being approved or rejected for credit.

https://archive.ics.uci.edu/ml/datasets/Credit+Approval

Data Set Information: This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

Attribute Information:
- A1: b, a.
- A2: continuous.
- A3: continuous.
- A4: u, y, l, t.
- A5: g, p, gg.
- A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- A7: v, h, bb, j, n, z, dd, ff, o.
- A8: continuous.
- A9: t, f.
- A10: t, f.
- A11: continuous.
- A12: t, f.
- A13: g, p, s.
- A14: continuous.
- A15: continuous.
- A16: +,- (class attribute)

Yes, most of that doesn't mean anything. A16 (the class attribute) is the most interesting, as it separates the 307 approved cases from the 383 rejected cases. The remaining variables have been obfuscated for privacy - a challenge you may have to deal with in your data science career.

Sprint challenges are evaluated based on satisfactory completion of each part. It is suggested you work through it in order, getting each aspect reasonably working, before trying to deeply explore, iterate, or refine any given step. Once you get to the end, if you want to go back and improve things, go for it!

## Part 1 - Load and validate the data

- Load the data as a `pandas` data frame.
- Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
- UCI says there should be missing data - check, and if necessary change the data so pandas recognizes it as na
- Make sure that the loaded features are of the types described above (continuous values should be treated as float), and correct as necessary

This is review, but skills that you'll use at the start of any data exploration. Further, you may have to do some investigation to figure out which file to load from - that is part of the puzzle.

In [17]:
#Load the data as a pandas data frame
import pandas as pd
credit_names = ["A" + str(i) for i in range(1,17)]
credit_data = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data', names = credit_names)
credit_data.head()

A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0   +

In [18]:
#Validate that it has the appropriate number of observations
credit_data.shape

(690, 16)

In [19]:
#Check for missing values. They are coded as ?s (There are no missing values in A16)
credit_data['A1'].value_counts()

b    468
a    210
?     12
Name: A1, dtype: int64

In [20]:
#code ?s as NaN
import numpy as np
credit_data = credit_data.replace("?", np.NaN)


In [21]:
#Make sure loaded features are of the types described above.
#A2, A11, A14, and A15 need to be fixed
credit_data.dtypes

A1      object
A2      object
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11      int64
A12     object
A13     object
A14     object
A15      int64
A16     object
dtype: object

In [33]:
#Change A2, A11, A14, and A15 to floats
fixcols = ['A2', 'A11', 'A14', 'A15']
credit_data[fixcols] = credit_data[fixcols].astype(float)
credit_data[fixcols].dtypes

A2     float64
A11    float64
A14    float64
A15    float64
dtype: object

## Part 2 - Exploring data, Testing hypotheses

The only thing we really know about this data is that A16 is the class label. Besides that, we have 6 continuous (float) features and 9 categorical features.

Explore the data: you can use whatever approach (tables, utility functions, visualizations) to get an impression of the distributions and relationships of the variables. In general, your goal is to understand how the features are different when grouped by the two class labels (`+` and `-`).

For the 6 continuous features, how are they different when split between the two class labels? Choose two features to run t-tests (again split by class label) - specifically, select one feature that is *extremely* different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

For the categorical features, explore by creating "cross tabs" (aka [contingency tables](https://en.wikipedia.org/wiki/Contingency_table)) between them and the class label, and apply the Chi-squared test to them. [pandas.crosstab](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html) can create contingency tables, and [scipy.stats.chi2_contingency](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) can calculate the Chi-squared statistic for them.

There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result (rejecting the null that the data are independent), and one where it is less extreme.

**NOTE** - "less extreme" just means smaller test statistic/larger p-value. Even the least extreme differences may be strongly statistically significant.

Your *main* goal is the hypothesis tests, so don't spend too much time on the exploration/visualization piece. That is just a means to an end - use simple visualizations, such as boxplots or a scatter matrix (both built in to pandas), to get a feel for the overall distribution of the variables.

This is challenging, so manage your time and aim for a baseline of at least running two t-tests and two Chi-squared tests before polishing. And don't forget to answer the questions in part 3, even if your results in this part aren't what you want them to be.

**For each test, assuming statistical significance as ($p>>.05$)**

$H_0: $ no significant difference between variable means.

$H_A: $ statistically significant difference between means.

In [34]:
credit_data.dtypes

A1      object
A2     float64
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11    float64
A12     object
A13     object
A14    float64
A15    float64
A16      int64
dtype: object

In [ ]:
#for the 6 continuous features, split into 2 data frames (approved and denied)
cont_var = ['A2', 'A3', 'A8', 'A11', 'A14', 'A15']
credit_data['A16'] = credit_data['A16'].replace({"+":1, "-":0})
approved = credit_data[credit_data['A16']==1]
denied = credit_data[credit_data['A16']==0]

#getting a weird error?????? but it did what I wanted so whatever. 

In [44]:
denied.head()

A1     A2   A3 A4 A5  A6  A7      A8 A9 A10  A11 A12 A13    A14  A15  A16
70  b  32.33  7.5  u  g   e  bb   1.585  t   f  0.0   t   s  420.0  0.0    0
71  b  34.83  4.0  u  g   d  bb  12.500  t   f  0.0   t   g    NaN  0.0    0
72  a  38.58  5.0  u  g  cc   v  13.500  t   f  0.0   t   g  980.0  0.0    0
73  b  44.25  0.5  u  g   m   v  10.750  t   f  0.0   f   s  400.0  0.0    0
74  b  44.83  7.0  y  p   c   v   1.625  f   f  0.0   f   g  160.0  2.0    0

In [45]:
approved.head()

A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14    A15  A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t  1.0   f   g  202.0    0.0    1
1  a  58.67  4.460  u  g  q  h  3.04  t   t  6.0   f   g   43.0  560.0    1
2  a  24.50  0.500  u  g  q  h  1.50  t   f  0.0   f   g  280.0  824.0    1
3  b  27.83  1.540  u  g  w  v  3.75  t   t  5.0   t   g  100.0    3.0    1
4  b  20.17  5.625  u  g  w  v  1.71  t   f  0.0   f   s  120.0    0.0    1

In [59]:
#graphs didn't help. idk how to tell which one is most/least significant without looking at ttest results. gonna run it on all of them
import scipy.stats as stats
for col in cont_var:
  results = stats.ttest_ind(approved[col], denied[col], nan_policy = 'omit')
  print("Results for " + col + ": " + str((results)))


Results for A2: Ttest_indResult(statistic=4.2922156166315535, pvalue=2.0276637071781407e-05)
Results for A3: Ttest_indResult(statistic=5.52998337614816, pvalue=4.551680702308068e-08)
Results for A8: Ttest_indResult(statistic=8.935819983773698, pvalue=3.6710537401601785e-18)
Results for A11: Ttest_indResult(statistic=11.667004222431277, pvalue=7.957718568079967e-29)
Results for A14: Ttest_indResult(statistic=-2.6358251986645476, pvalue=0.008586135473979569)
Results for A15: Ttest_indResult(statistic=4.680216020964486, pvalue=3.4520256956287944e-06)


**The most extreme statistically significant difference is in A11, with T=11.67 and pvalue=7.96e-29.
The least extreme is A14, with T=2.64 and a pvalue of .008 (assuming significance at pvalue=.01)**

In [60]:
cat_var = ['A1', 'A4', 'A5', 'A6', 'A7', 'A9', 'A10', 'A12', 'A13']
p = []
for col in cat_var:
  
  ctab = pd.crosstab(credit_data[col], credit_data['A16'])
  print(col +" against A16: \n") #puts a title on the crosstab to know which cell
  print(ctab)
    
  chi_squared, p_value, dof, expected = stats.chi2_contingency(ctab)
  print("\n")
  p.append(p_value)
  print(f"Chi-Squared: {chi_squared}")
  print(f"P-value: {p_value}")

  if(p_value < 0.01):
    print("Significant with p < 0.01!")
    
  print(f"Degrees of Freedom: {dof}") 
  print("Expected: \n", np.array(expected))
  print("\n") #figured out how to add line breaks. more readable. 
print(p)

A1 against A16: 

A16    0    1
A1           
a    112   98
b    262  206


Chi-Squared: 0.3112832649161995
P-value: 0.5768937883001117
Degrees of Freedom: 1
Expected: 
 [[115.84070796  94.15929204]
 [258.15929204 209.84070796]]
A4 against A16: 

A16    0    1
A4           
l      0    2
u    263  256
y    118   45


Chi-Squared: 26.234074966202144
P-value: 2.010680204180363e-06
Significant with p < 0.01!
Degrees of Freedom: 2
Expected: 
 [[  1.11403509   0.88596491]
 [289.09210526 229.90789474]
 [ 90.79385965  72.20614035]]
A5 against A16: 

A16    0    1
A5           
g    263  256
gg     0    2
p    118   45


Chi-Squared: 26.234074966202144
P-value: 2.010680204180363e-06
Significant with p < 0.01!
Degrees of Freedom: 2
Expected: 
 [[289.09210526 229.90789474]
 [  1.11403509   0.88596491]
 [ 90.79385965  72.20614035]]
A6 against A16: 

A16   0   1
A6         
aa   35  19
c    75  62
cc   12  29
d    23   7
e    11  14
ff   46   7
i    45  14
j     7   3
k    37  14
m    22  16
q    

**The smallest p value occured for `A9`: a value of $~3.116*10^{-79}$, with a $x^2$ value of $~355.2$.

**The differences for `A1` and `A12` were not statistically signifcant ($p>>0.01$).

**`A13` was also not statistically signifiant at ($p>>.01$), but was significant if pvalue is adjusted to ($p>.05$).

## Part 3 - Analysis and Interpretation

Now that you've looked at the data, answer the following questions:

- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?
- Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?
- What was the most challenging part of this sprint challenge?

Answer with text, but feel free to intersperse example code/results or refer to it from earlier.

**Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?**

Each of the continuous variables were statistically significant to some degree, with A11 having the highest significance and A14 having the lowest. This indicates that each of these variables have some direct influence in determining who is approved and who is denied credit. A lack of statistical significance would have meant a variable was evenly distributed across those who were approved, and those who were denied. 

**Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?**

Some of the categorical variables had extremely low p-values, which suggests that they are extremely influential in the application process. These probably relate to relevant qualitative factors, such as previous loan delinquency.**

A1 and A12 had high pvalues, indicating that the may not directly influence credit approval, though still have some influence. Though A13 barely missed the cutoff for significance at p > .01, it is significant at p>.05, suggesting that it may be directly related to approval status, but to a lesser degree than the other statistically significant variables.

All other categorical variables appear to be directly related to approval status based on very low pvalues.

**What was the most challenging part of this sprint challenge?**

I spent a lot of time trying to decide which two columns to run t test and which two to run Chi-squared test in based on graphs, means, etc. before I decided to just test them all and look at the results that way. Once that was out of the way, interpretation was straightforward. 